### ISSUE TITLE
Parsing of packets

### ISSUE THREAD

**Issue by lal12** [tediousjs/tedious#936]
Hello,
I need to parse TDS packets myself. As far as I have seen there isn't an api for that built in? Could someone give me a hint where in the source code the parsing "starts" = where the packet buffer is passed to function to be converted to a native datatype...

**Comment by arthurschreiber** [comment]
You can access the functionality you need, it's just not something that's really exposed. You'll have to not require the module via `require('tedious')`, but the individual functionality via `require('tedious/lib/...')`

There is an `IncomingMessageStream` class that you can load via `require('tedious/lib/incoming-message-stream')` that is a stream that will parse raw TDS data into individual TDS messages. TDS messages are represented by individual `Message` instances, which themselves are streams again of the data contained in the message.

Do you also need to parse the actual tokens inside the message? Where are you getting your data from? Do you also need to write messages?

**Comment by lal12** [comment]
Thanks for the quick reply. I am actually developing a small proxy. I have a CAD software which requires a specific Versions and Editions of MSSQL Server, but as far as I can tell, not really for a technical reason. 

And I don't really wanna buy the full edition of MSSQL, since the actual use case is ridiculous simple and while all the features of a full blown Edition might be worth the price, none of them are actually used. So currently doing this for more like a proof of concept, if it works well maybe also in production. So much about the backstory, probably more a whining about the stupid program, which is expensive and really bad programmed in general ^^

So I want to fake the version information and potentially some other packets send by the server. So yes I need to parse the tokens in a server table response and would need to also write packets, however I might prefer to just replace the relevant bytes in the buffer for better speed, instead of re generating the whole packet.

**Comment by arthurschreiber** [comment]
@lal12 Building a proxy that does this should be possible, pretty straightforward even! 🙇Please let us know if you get this working, or if you run into any troubles doing so! ❤️

**Comment by lal12** [comment]
Well I actually got it to work. I had to reimplement some of tedious' classes a bit different for me, but for all of the parsing and a part of the writing/generating I could take tedious' code.

Currently I am replacing the values in the table responses. It probably would be easier and better to just replaces specific parts of the SQL Query, in which the manipulation would also work in more complicated queries, but since it is a proof of concept, doing it by manipulating the responses seemed to be a good idea.

I will clean up the project and will post the github repo here, when created. 

Thanks for the help again.
